cli/core/multicall/multicall.go

```diff
@@ -0,0 +1,229 @@
+	callData, err := abi.Pack(method, params...)
+	if err != nil {
+		return nil, fmt.Errorf("error packing multicall: %s", err.Error())
+	}
+	return &MultiCallMetaData[T]{
+		Address:     contractAddress,
+		Data:        callData,
+		Deserialize: deserialize,
+	}, nil
+}
+
+func DoMultiCall[A any, B any](mc MulticallContract, a *MultiCallMetaData[A], b *MultiCallMetaData[B]) (*A, *B, error) {
+	res, err := doMultiCallMany(mc, a.Raw(), b.Raw())
+	if err != nil {
+		return nil, nil, fmt.Errorf("error performing multicall: %s", err.Error())
+	}
+	return any(res[0].Value).(*A), any(res[1].Value).(*B), nil
+}
+
+func DoMultiCallMany[A any](mc MulticallContract, requests ...*MultiCallMetaData[A]) (*[]A, error) {
+	res, err := doMultiCallMany(mc, utils.Map(requests, func(mc *MultiCallMetaData[A], index uint64) RawMulticall {
+		return mc.Raw()
+	})...)
+	if err != nil {
+		return nil, fmt.Errorf("multicall failed: %s", err.Error())
+	}
+
+	// unwind results
+	unwoundResults := utils.Map(res, func(d DeserializedMulticall3Result, i uint64) A {
+		// force these back to A
+		return any(d.Value).(A)
+	})
+	return &unwoundResults, nil
+}
+
+/*
+ * Some RPC providers may limit the amount of calldata you can send in one eth_call, which (for those who have 1000's of validators), means
+ * you can't just spam one enormous multicall request.
+ *
+ * This function checks whether the calldata appended exceeds maxBatchSizeBytes
+ */
+func chunkCalls(allCalls []ParamMulticall3Call3, maxBatchSizeBytes int) [][]ParamMulticall3Call3 {
+	// chunk by the maximum size of calldata, which is 1024 per call.
+	results := [][]ParamMulticall3Call3{}
+	currentBatchSize := 0
+	currentBatch := []ParamMulticall3Call3{}
+
+	for _, call := range allCalls {
+		if (currentBatchSize + len(call.CallData)) > maxBatchSizeBytes {
+			// we can't fit in this batch, so dump the current batch and start a new one
+			results = append(results, currentBatch)
+			currentBatchSize = 0
+			currentBatch = []ParamMulticall3Call3{}
+		}
+
+		currentBatch = append(currentBatch, call)
+		currentBatchSize += len(call.CallData)
+	}
+
+	// check if we forgot to add the last batch
+	if len(currentBatch) > 0 {
+		results = append(results, currentBatch)
+	}
+
+	return results
+}
+
+func doMultiCallMany(mc MulticallContract, calls ...RawMulticall) ([]DeserializedMulticall3Result, error) {
+	typedCalls := make([]ParamMulticall3Call3, len(calls))
+	for i, call := range calls {
+		typedCalls[i] = ParamMulticall3Call3{
+			Target:       call.Address,
+			AllowFailure: true,
+			CallData:     call.Data,
+		}
+	}
+
+	// see if we need to chunk them now
+	chunkedCalls := chunkCalls(typedCalls, func() int {
+		if mc.MaxBatchSize == 0 {
+			return math.MaxInt64
+		} else {
+			return int(mc.MaxBatchSize)
+		}
+	}())
+	var results = make([]interface{}, len(calls))
+	var totalResults = 0
```

Comment:
hm, this seems race conditiony. probably want to implement this with goroutines and mutexes

Reply:
sorry this is isn't concurrent. do we care?

Reply:
can add incrementally in the future

Reply:
should be no race conditions, as this is entirely single threaded :p